dialogue-client-verifier/src/test/java/com/palantir/verification/BinaryReturnTypeTest.java

```diff
@@ -0,0 +1,227 @@
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.zip.GZIPOutputStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class BinaryReturnTypeTest {
+    private static final UserAgent USER_AGENT = UserAgent.of(UserAgent.Agent.of("BinaryReturnTypeTest", "0.0.0"));
+    private static final SslConfiguration SSL_CONFIG = SslConfiguration.of(
+            Paths.get("../dialogue-core/src/test/resources/trustStore.jks"),
+            Paths.get("../dialogue-core/src/test/resources/keyStore.jks"),
+            "keystore");
+
+    private Undertow undertow;
+    private HttpHandler undertowHandler;
+
+    @Before
+    public void before() {
+        undertow = Undertow.builder()
+                .addHttpListener(
+                        0, "localhost", new BlockingHandler(exchange -> undertowHandler.handleRequest(exchange)))
+                .build();
+        undertow.start();
+    }
+
+    @Test
+    public void conjure_generated_async_interface_with_optional_binary_return_type_and_gzip() {
+        setBinaryGzipResponse("Hello, world");
+        SampleServiceAsync client = sampleServiceAsync();
+
+        ListenableFuture<Optional<InputStream>> future = client.getOptionalBinary();
+        Optional<InputStream> maybeBinary = Futures.getUnchecked(future);
+
+        assertThat(maybeBinary).isPresent();
+        assertThat(maybeBinary.get()).hasSameContentAs(asInputStream("Hello, world"));
+    }
+
+    @Test
+    public void conjure_generated_blocking_interface_with_optional_binary_return_type_and_gzip() {
+        setBinaryGzipResponse("Hello, world");
+
+        Optional<InputStream> maybeBinary = sampleServiceBlocking().getOptionalBinary();
+
+        assertThat(maybeBinary).isPresent();
+        assertThat(maybeBinary.get()).hasSameContentAs(asInputStream("Hello, world"));
+    }
+
+    @Test
+    public void stream_3_gigabytes() throws IOException {
+        long oneMegabyte = 1000_000;
+        int megabytes = 3000;
+        long limit = megabytes * oneMegabyte;
+        assertThat(limit).isGreaterThan(Integer.MAX_VALUE);
+
+        byte[] sample = new byte[8192];
+        Arrays.fill(sample, (byte) 'A');
+
+        undertowHandler = exchange -> {
+            exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/octet-stream");
+            InputStream bigInputStream = repeat(sample, limit);
+            ByteStreams.copy(bigInputStream, exchange.getOutputStream());
+        };
+
+        Stopwatch sw = Stopwatch.createStarted();
+
+        InputStream maybeBinary = sampleServiceBlocking().getOptionalBinary().get();
+        assertThat(ByteStreams.exhaust(maybeBinary))
+                .describedAs("Should receive exactly the number of bytes we sent!")
+                .isEqualTo(limit);
+
+        System.out.printf("%d MB took %d millis%n", megabytes, sw.elapsed(TimeUnit.MILLISECONDS));
```

Comment:
maybe it's just my mac, but this takes like 28 seconds for 3GB sometimes

Reply:
That’s odd considering we’re using plain text http, how long does the build system take?